test/dynamo/test_higher_order_ops.py

```diff
@@ -2501,6 +2502,75 @@ def fn(pred, pytree_in):
             ):
                 torch.compile(fn, backend="eager")(pred, pytree_in)
 
+    def test_hints_wrapper(self):
+        def ref_fn(x, y):
+            x = x + y
+            x = torch.relu(x)
+            x = x + y
+            return torch.abs(x)
+
+        def fn_with_hints(x, y):
+            x = x + y
+
+            def inner_body_fn(x, y):
+                x = torch.relu(x)
+                x = x + y
+                return x
+
+            def outer_body_fn(x, y):
+                x = hints_wrapper(inner_body_fn, (x, y), {}, hints={"inner_body": True})
+                x = torch.abs(x)
+                return x
+
+            res = hints_wrapper(outer_body_fn, (x, y), {}, hints={"outer_body": True})
+            return res
+
+        backend = EagerAndRecordGraphs()
+        cnt = CompileCounterWithBackend(backend)
+
+        x = torch.randn(2, 4)
+        y = torch.ones(4)
+
+        aot_eager_res = torch.compile(fn_with_hints, backend="aot_eager")(x, y)
+        compiled_res = torch.compile(fn_with_hints, backend=cnt)(x, y)
+        ref_res = ref_fn(x, y)
+        self.assertEqual(compiled_res, ref_res)
+        self.assertEqual(aot_eager_res, ref_res)
+        self.assertEqual(len(cnt.graphs), 1)
+        graph = backend.graphs[0]
+        self.assertExpectedInline(
+            normalize_gm(graph.print_readable(print_output=False)),
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self, L_x_: "f32[2, 4]", L_y_: "f32[4]"):
+        l_x_ = L_x_
+        l_y_ = L_y_
+
+        x: "f32[2, 4]" = l_x_ + l_y_;  l_x_ = None
+
+        hints_wrapper_body_1 = self.hints_wrapper_body_1
+        hints_wrapper = torch.ops.higher_order.hints_wrapper(hints_wrapper_body_1, (x, l_y_), {}, hints = {'outer_body': True});  hints_wrapper_body_1 = x = l_y_ = None
+        res: "f32[2, 4]" = hints_wrapper[0];  hints_wrapper = None
+        return (res,)
+
+    class hints_wrapper_body_1(torch.nn.Module):
+        def forward(self, x: "f32[2, 4]", l_y_: "f32[4]"):
+            hints_wrapper_body_0 = self.hints_wrapper_body_0
+            hints_wrapper = torch.ops.higher_order.hints_wrapper(hints_wrapper_body_0, (x, l_y_), {}, hints = {'inner_body': True});  hints_wrapper_body_0 = x = l_y_ = None
+            x_1: "f32[2, 4]" = hints_wrapper[0];  hints_wrapper = None
+
+            x_2: "f32[2, 4]" = torch.abs(x_1);  x_1 = None
+            return (x_2,)
+
+        class hints_wrapper_body_0(torch.nn.Module):
+            def forward(self, x: "f32[2, 4]", l_y_: "f32[4]"):
+                x_1: "f32[2, 4]" = torch.relu(x);  x = None
+
+                x_2: "f32[2, 4]" = x_1 + l_y_;  x_1 = l_y_ = None
+                return (x_2,)
+""",
+        )
+
```

Comment:
A few more thoughts:
1. Do we want to support the case when users not going through torch.compile (like directly calling the operator from eager without calling torch.compile)? If we do, we probably should add a test case for this. Or error out early if we don't.
2. what will happen if there's a graph break inside the subgraphs e.g. a print statement
3. Can add an export test in test/export/test_export.py?

Reply:
> Do we want to support the case when users not going through torch.compile (like directly calling the operator from eager without calling torch.compile)? If we do, we probably should add a test case for this. Or error out early if we don't.

Yes, we support eager mode. Please check test.

> what will happen if there's a graph break inside the subgraphs e.g. a print statement

We will error out if there is graph break, following same way for torch.cond.

> Can add an export test in test/export/test_export.py?

I tried to add a test in test_export.py and got below graph. Does it mean it can work with export?

```python
ExportedProgram:
    class GraphModule(torch.nn.Module):
        def forward(self, x: "f32[2, 4]", y: "f32[4]"):
             # File: /home/wzhang/workspace/pytorch-upstream/test/export/test_export.py:6723 in forward, code: x = x + y
            add: "f32[2, 4]" = torch.ops.aten.add.Tensor(x, y);  x = None

             # File: /home/wzhang/workspace/pytorch-upstream/test/export/test_export.py:6735 in forward, code: res = hints_wrapper(outer_body_fn, (x, y), {}, hints={"outer_body": True})
            hints_wrapper_body_graph_0 = self.hints_wrapper_body_graph_0
            hints_wrapper = torch.ops.higher_order.hints_wrapper(hints_wrapper_body_graph_0, (add, y), {}, hints = {'outer_body': True});  hints_wrapper_body_graph_0 = add = y = None
            getitem: "f32[2, 4]" = hints_wrapper[0];  hints_wrapper = None
            return (getitem,)

        class hints_wrapper_body_graph_0(torch.nn.Module):
            def forward(self, arg0_1: "f32[2, 4]", arg1_1: "f32[4]"):
                 # File: /home/wzhang/workspace/pytorch-upstream/test/export/test_export.py:6735 in forward, code: res = hints_wrapper(outer_body_fn, (x, y), {}, hints={"outer_body": True})
                hints_wrapper_body_graph_0 = self.hints_wrapper_body_graph_0
                hints_wrapper = torch.ops.higher_order.hints_wrapper(hints_wrapper_body_graph_0, (arg0_1, arg1_1), {}, hints = {'inner_body': True});  hints_wrapper_body_graph_0 = arg0_1 = arg1_1 = None
                getitem: "f32[2, 4]" = hints_wrapper[0];  hints_wrapper = None
                abs_1: "f32[2, 4]" = torch.ops.aten.abs.default(getitem);  getitem = None
                return (abs_1,)

            class hints_wrapper_body_graph_0(torch.nn.Module):
                def forward(self, arg0_1: "f32[2, 4]", arg1_1: "f32[4]"):
                     # File: /home/wzhang/workspace/pytorch-upstream/test/export/test_export.py:6735 in forward, code: res = hints_wrapper(outer_body_fn, (x, y), {}, hints={"outer_body": True})
                    relu: "f32[2, 4]" = torch.ops.aten.relu.default(arg0_1);  arg0_1 = None
                    add: "f32[2, 4]" = torch.ops.aten.add.Tensor(relu, arg1_1);  relu = arg1_1 = None
                    return (add,)

Graph signature: ExportGraphSignature(input_specs=[InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='x'), target=None, persistent=None), InputSpec(kind=<InputKind.USER_INPUT: 1>, arg=TensorArgument(name='y'), target=None, persistent=None)], output_specs=[OutputSpec(kind=<OutputKind.USER_OUTPUT: 1>, arg=TensorArgument(name='getitem'), target=None)])
Range constraints: {}
```

Reply:
yeah, it exports

Reply:
Thanks. I just did another updates according to your comments. Please take another look.